WordPress/gutenberg
# Fix resolver resolution status key types mismatch causing unwanted resolver calls for identical state data

## What?


Stops unwanted resolver calls due to resolver cache misses. 

Specifically fixes `getEntityRecord` in `@wordpress/core-data` to avoid cache miss caused by type mismatch on the `recordKey` argument.

Closes https://github.com/WordPress/gutenberg/issues/52106

## Why?


The following selector calls are not considered equivalent by the caching mechanism in `@wordpress/data` (see https://github.com/WordPress/gutenberg/issues/52106 for detailed explanation):

```js
getEntityRecord( 'postType', 'post', 1 ); // number as recordKey
getEntityRecord( 'postType', 'post', '1' ); // string as recordKey
```

This means that even if the record exists in state, the relevant _resolver_ is not marked as resolved and thus the resolver runs again thereby triggering a network request.

This is very likely to occur because in Gutenberg `ID`s can be strings which means a lot of (potential) cache misses and unnecessary network calls and loading delays.

A specific example is the `useNavigator` hooks where the resulting `params` object will contain params as strings even if they are numeric. So if there's a param of `postId` then it will be a string. It's very easy to accidentally pass that to `getEntityRecord`.


## How?


Adds a new optional property to the ~resolver~ selector which allows the implementor to define a method which normalizes the ~resolver~ `args` passed to the selector into to a standard form. This standardised form is then used in cache set/get.

This PR implements such a method for the `getEntityRecord` ~resolver~ selector to stabize the type of the `recordKey` as a number rather than a string when the `recordKey` argument looks like a numberic ID (which is likely to be the ID of a Post record).

## Testing Instructions





- Create some Navigation Menus on your website using the Navigation block. Be sure to save/publish them.
- Create a new Post
- Open dev tools network tab and then hit `Esc` to _also_ open the `Console` tab below it.
- In console type `wp.data.select('core').getEntityRecords('postType', 'wp_navigation')`. You should see `null` and then running it _again_ you should see results. This has populated the `core/data` state with entities of that type.
- Note down the `id`s of those records.
- At this point those records are in state and thus any calls to select the individual records should _not_ result in a network request because they are already in state and `getEntityRecords` [resolves the `getEntityRecord` selectors](https://github.com/WordPress/gutenberg/blob/3b7902e6885fd12c660e4cfd94f115ebfbe91434/packages/core-data/src/resolvers.js#L199-L217). 
- Choose the ID of one of your records and enter the following into the console making sure that you pass the 3rd argument as a _string_ and _not_ a int value - `wp.data.select('core').getEntityRecord('postType', 'wp_navigation','{YOUR_RECORD_ID}')`
- You should _**not**_ see `GET` request (ignore `OPTIONS`) to `wp/v2/navigation/{RECORD_ID}`.
- Make the same request but this time make the `recordKey` argument a `Number`.
- See no network request is dispatched.

...you can also test this in the Site Editor by:

- Clicking `Navigation` in Browse Mode sidebar.
- Clicking on one of your Navigations.
- See that no `GET` network request was dispatched.

### Testing Instructions for Keyboard


## Screenshots or screencast
